From 6d4ab1572f04be8104f22ddb19504ec1fbe4be33 Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Mon, 16 Mar 2026 21:03:19 +0800
Subject: [PATCH 26/71] video: rockchip: rga3: fix return value of
 rga_mm_lookup_iova() on error

Change-Id: I2c5d066fffb2e72e9b5ba3affa44160758b3069e
Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
(cherry picked from commit df53ec9ff0bfb96e86756f6b601b555a709d4faa)
---
 drivers/video/rockchip/rga3/rga_mm.c | 22 +++++++++++++++-------
 1 file changed, 15 insertions(+), 7 deletions(-)

diff --git a/drivers/video/rockchip/rga3/rga_mm.c b/drivers/video/rockchip/rga3/rga_mm.c
index 88623fca782dc..1994ef25aaae7 100644
--- a/drivers/video/rockchip/rga3/rga_mm.c
+++ b/drivers/video/rockchip/rga3/rga_mm.c
@@ -1236,7 +1236,7 @@ int rga_mm_lookup_flag(struct rga_mm *mm_session, uint64_t handle)
 dma_addr_t rga_mm_lookup_iova(struct rga_internal_buffer *buffer)
 {
 	if (rga_mm_is_invalid_dma_buffer(buffer->dma_buffer))
-		return 0;
+		return DMA_MAPPING_ERROR;
 
 	return buffer->dma_buffer->iova + buffer->dma_buffer->offset;
 }
@@ -1639,13 +1639,15 @@ static int rga_mm_sync_dma_sg_for_device(struct rga_internal_buffer *buffer,
 
 	if (buffer->mm_flag & RGA_MEM_PHYSICAL_CONTIGUOUS) {
 		if (scheduler->data->mmu == RGA_IOMMU) {
-			if (rga_mm_is_invalid_dma_buffer(buffer->dma_buffer)) {
-				rga_job_err(job, "invalid dma-buffer with IOMMU device!\n");
+			dma_addr_t iova = rga_mm_lookup_iova(buffer);
+
+			if (dma_mapping_error(buffer->dma_buffer->map_dev, iova)) {
+				rga_job_err(job, "invalid iova for dma-buffer with IOMMU device!\n");
 				return -EFAULT;
 			}
 
 			dma_sync_single_for_device(buffer->dma_buffer->map_dev,
-				buffer->dma_buffer->iova, buffer->dma_buffer->size, dir);
+				iova, buffer->size, dir);
 		} else {
 			dma_sync_single_for_device(scheduler->dev,
 				buffer->phys_addr, buffer->size, dir);
@@ -1693,13 +1695,15 @@ static int rga_mm_sync_dma_sg_for_cpu(struct rga_internal_buffer *buffer,
 
 	if (buffer->mm_flag & RGA_MEM_PHYSICAL_CONTIGUOUS) {
 		if (scheduler->data->mmu == RGA_IOMMU) {
-			if (rga_mm_is_invalid_dma_buffer(buffer->dma_buffer)) {
-				rga_job_err(job, "invalid dma-buffer with IOMMU device!\n");
+			dma_addr_t iova = rga_mm_lookup_iova(buffer);
+
+			if (dma_mapping_error(buffer->dma_buffer->map_dev, iova)) {
+				rga_job_err(job, "invalid iova for dma-buffer with IOMMU device!\n");
 				return -EFAULT;
 			}
 
 			dma_sync_single_for_cpu(buffer->dma_buffer->map_dev,
-				buffer->dma_buffer->iova, buffer->dma_buffer->size, dir);
+				iova, buffer->size, dir);
 		} else {
 			dma_sync_single_for_cpu(scheduler->dev,
 				buffer->phys_addr, buffer->size, dir);
@@ -1748,6 +1752,10 @@ static int rga_mm_get_buffer_info(struct rga_job *job,
 		}
 
 		addr = rga_mm_lookup_iova(internal_buffer);
+		if (dma_mapping_error(internal_buffer->dma_buffer->map_dev, (dma_addr_t)addr)) {
+			rga_job_err(job, "invalid iova for dma-buffer with IOMMU device!\n");
+			return -EFAULT;
+		}
 
 		break;
 	case RGA_MMU:
